Add torch.unflatten and improve its docs

fdrocha opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81231
* #81230
* __->__ #81399

    unflatten now has a free function version in torch.flatten in addition to
    the method in torch.Tensor.flatten.

    Updated docs to reflect this and polished them a little.
    For consistency, changed the signature of the int version of unflatten in
    native_functions.yaml.

    Some override tests were failing because unflatten has unusual
    characteristics in terms of the .int and .Dimname versions having
    different number of arguments so this required some changes
    to test/test_override.py

    Removed support for using mix of integer and string arguments
    when specifying dimensions in unflatten.

Commit:
unflatten now has a free function version
in torch.flatten in addition to the method
in torch.Tensor.flatten.

Updated docs to reflect this and polished
them a little.

[ghstack-poisoned]

Commit:
Update on "Add torch.unflatten and improve its docs"


in torch.flatten in addition to the method
in torch.Tensor.flatten.

Updated docs to reflect this and polished
them a little.

[ghstack-poisoned]

Commit:
Update on "Add torch.unflatten and improve its docs"


in torch.flatten in addition to the method
in torch.Tensor.flatten.

Updated docs to reflect this and polished
them a little.

[ghstack-poisoned]

Lezcano commented on test/test_overrides.py:
```diff
@@ -649,7 +649,11 @@ def instance_gen():
                 elif t == 'bool':
                     func_args.append(False)
-                elif t.startswith('int') or t in {'Dimname', 'DimnameList'}:
+                elif t == 'Dimname':
+                    func_args.append("")
+                elif t == 'DimnameList':
+                    func_args.append([""])
+                elif t.startswith('int'):
                     func_args.append(0)
```
@peterbell10 does this change sound reasonable?

peterbell10 replied:
Seems reasonable to me, a dimname should be a string in the python api.

Commit:
Update on "Add torch.unflatten and improve its docs"


in torch.flatten in addition to the method
in torch.Tensor.flatten.

Updated docs to reflect this and polished
them a little.

[ghstack-poisoned]

Commit:
Update on "Add torch.unflatten and improve its docs"


in torch.flatten in addition to the method
in torch.Tensor.flatten.

Updated docs to reflect this and polished
them a little.

[ghstack-poisoned]

Lezcano commented on torch/_torch_docs.py:
```diff
@@ -4050,6 +4050,35 @@ def merge_dicts(*dicts):
+
+Expands a dimension of the input tensor over multiple dimensions.
+
+A tensor reshaping function that does the opposite of :func:`torch.flatten`.
```
What about using `seealso`?
```suggestion
.. seealso::

    :func:`torch.flatten` the inverse of this function. It coalesces several dimensions into one.
```

Lezcano commented on torch/_torch_docs.py:
```diff
@@ -4050,6 +4050,35 @@ def merge_dicts(*dicts):
+         equal ``input.shape[dim]``.
+
+Returns:
+    A ``View`` of ``input`` with the specified dimension unflattened.
```
```suggestion
    A view of the input with the specified dimension unflattened.
```

Lezcano commented on torch/_torch_docs.py:
```diff
@@ -4050,6 +4050,35 @@ def merge_dicts(*dicts):
+Returns:
+    A ``View`` of ``input`` with the specified dimension unflattened.
+
+Examples:
```
I don't know why the examples are being rendered funny in the docs. I think you may need to write
```suggestion
Examples::

```
See how it's rendered at the moment (taken from the message from facebook-github-bot) https://docs-preview.pytorch.org/81399/generated/torch.unflatten.html?highlight=unflatten#torch.unflatten

Lezcano commented on torch/_torch_docs.py:
```diff
@@ -4050,6 +4050,35 @@ def merge_dicts(*dicts):
+    torch.Size([3, 2, 2, 1])
+    >>> torch.unflatten(torch.randn(3, 4, 1), 1, (-1, 2)).shape
+    torch.Size([3, 2, 2, 1])
+    >>> torch.unflatten(torch.randn(5, 12, 3), -1, (2,2,3,1,1)).shape
```
```suggestion
    >>> torch.unflatten(torch.randn(5, 12, 3), -1, (2, 2, 3, 1, 1)).shape
```

lezcano reviewed (commented):
A few minor things, but otherwise looks great! Thank you Fabio!

fdrocha commented on test/test_namedtensor.py:
```diff
@@ -1083,22 +1083,13 @@ def test_flatten_nodims(self):
+            torch.ones(4, names=('A',)).unflatten('A', [('A', 2), ('B', 2)]),
             torch.ones(2, 2, names=('A', 'B'))))
         self.assertTrue(torch.equal(
-            torch.ones(4).unflatten(0, (['A', 2], ['B', 2])),
-            torch.ones(2, 2, names=('A', 'B'))))
-        self.assertTrue(torch.equal(
-            torch.ones(4).unflatten(-1, (['A', 2], ['B', 2])),
-            torch.ones(2, 2, names=('A', 'B'))))
-        self.assertTrue(torch.equal(
-            torch.ones(4).unflatten(-1, (['A', -1], ['B', 2])),
-            torch.ones(2, 2, names=('A', 'B'))))
-        self.assertTrue(torch.equal(
-            torch.ones(4).unflatten(-1, (['A', 2], ['B', -1])),
```
Like we discussed, I removed the overload of unflatten that allows you to do stuff like `t.unflatten(0, (['A', -1], ['B',2]))`. You have to do something like `t.unflatten('A', (['A', -1], ['B',2]))` instead. 

This meant that the tests in here had to be changed, and after making the changes there were duplicate tests and I removed them (i.e., `t.unflatten(0, (['A', -1], ['B',2])` and `t.unflatten(-1, (['A', -1], ['B',2])` both became `t.unflatten('A', (['A', -1], ['B',2])` ).

fdrocha committed:
Update torch/_torch_docs.py

Co-authored-by: Mario Lezcano Casado <Lezcano@users.noreply.github.com>

fdrocha committed:
Update torch/_torch_docs.py

Co-authored-by: Mario Lezcano Casado <Lezcano@users.noreply.github.com>

fdrocha committed:
Update torch/_torch_docs.py

Co-authored-by: Mario Lezcano Casado <Lezcano@users.noreply.github.com>

fdrocha committed:
Update torch/_torch_docs.py

Co-authored-by: Mario Lezcano Casado <Lezcano@users.noreply.github.com>

lezcano reviewed (approved):
LGTM

Commit:
Update on "Add torch.unflatten and improve its docs"


    unflatten now has a free function version in torch.flatten in addition to
    the method in torch.Tensor.flatten.

    Updated docs to reflect this and polished them a little.
    For consistency, changed the signature of the int version of unflatten in
    native_functions.yaml.

    Some override tests were failing because unflatten has unusual
    characteristics in terms of the .int and .Dimname versions having
    different number of arguments so this required some changes
    to test/test_override.py

    Removed support for using mix of integer and string arguments
    when specifying dimensions in unflatten.

[ghstack-poisoned]

Commit:
Update on "Add torch.unflatten and improve its docs"


    unflatten now has a free function version in torch.flatten in addition to
    the method in torch.Tensor.flatten.

    Updated docs to reflect this and polished them a little.
    For consistency, changed the signature of the int version of unflatten in
    native_functions.yaml.

    Some override tests were failing because unflatten has unusual
    characteristics in terms of the .int and .Dimname versions having
    different number of arguments so this required some changes
    to test/test_override.py

    Removed support for using mix of integer and string arguments
    when specifying dimensions in unflatten.

[ghstack-poisoned]

ngimel commented:
@pytorchbot merge

ngimel commented:
@pytorchbot rebase

ngimel commented:
@fdrocha can you please manually rebase and land?

Commit:
Update on "Add torch.unflatten and improve its docs"


    unflatten now has a free function version in torch.flatten in addition to
    the method in torch.Tensor.flatten.

    Updated docs to reflect this and polished them a little.
    For consistency, changed the signature of the int version of unflatten in
    native_functions.yaml.

    Some override tests were failing because unflatten has unusual
    characteristics in terms of the .int and .Dimname versions having
    different number of arguments so this required some changes
    to test/test_override.py

    Removed support for using mix of integer and string arguments
    when specifying dimensions in unflatten.

[ghstack-poisoned]

fdrocha commented:
@pytorchbot merge -g

Commit:
Update on "Add torch.unflatten and improve its docs"


    unflatten now has a free function version in torch.flatten in addition to
    the method in torch.Tensor.flatten.

    Updated docs to reflect this and polished them a little.
    For consistency, changed the signature of the int version of unflatten in
    native_functions.yaml.

    Some override tests were failing because unflatten has unusual
    characteristics in terms of the .int and .Dimname versions having
    different number of arguments so this required some changes
    to test/test_override.py

    Removed support for using mix of integer and string arguments
    when specifying dimensions in unflatten.

[ghstack-poisoned]

fdrocha commented:
@pytorchbot merge -g

fdrocha commented:
This is now failing with

```
The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 

Broken ops: [
	aten::unflatten.int(Tensor(a) self, int dim, int[] sizes, str[]? names=None) -> Tensor(a)
]
```

(Not sure why it wasn't hitting this error before, I guess the tests must have been improved since CI last ran)

This change was intentional and suggested by @Lezcano. The reasoning was that signature was very unusual and it seemed better to replace it with something more regular. Also, the unusual signature made some override tests fail when the free function variant was added. 

The new signature is `(Tensor(a) self, int dim, int[] sizes)` (there's also a pure Dimname variant). 
This means that code like

```
torch.randn(2).unflatten(0, (('B1', -1), ('B2', 1)))
```

will no longer work, but that seems like inconsistent use of Dimnames and ints for specifying dimensions and not clear it should be supported. Note that something like

```
torch.randn(2, names=('A',)).unflatten('A', (('B1', -1), ('B2', 1)))
```

will still work.

Finally, this only affects deprecated Dimname functionality.

@ngimel WDYT?

ngimel commented:
If we can make this without BC breaking changes we should. While I'm ok removing mixed dimname and number indexing, the complaint is about regular overload, aten::unflatten.int, not the mixed one.

fdrocha commented:
> If we can make this without BC breaking changes we should. 
We can, we just have to skip some overload tests.
I think main cost is we will end up with a unique type of signature for `torch.unflatten` that doesn't appear anywhere else and might require some special casing in tests.

> While I'm ok removing mixed dimname and number indexing, the complaint is about regular overload, aten::unflatten.int, not the mixed one.

The complaint is about 

```
aten::unflatten.int(Tensor(a) self, int dim, int[] sizes, str[]? names=None) -> Tensor(a)
```

which _is_ the mixed one. If you don't provide the last argument `names` then it's pure int dimensions, the proposed change is removing that last optional argument from the signature so it's no longer mixed.

Anyway, happy to do it either way, let me know what you prefer @ngimel

Lezcano commented:
This is expected.  That use case was very weird (mixing numbers and dimnames) and it was the only op in PyTorch that had it. Keeping the support for this would have lead to an unnecessarily complicated code, so we dedided to remove it.
Just add it to the relevant skip list.

Fwiw, @ezyang has an open PR trying to remove all dimnames from PyTorch (I really hope he manages to get it merged), so removing thisbparticularly odd overload is not that big a deal IMO

fdrocha commented:
> This is expected. That use case was very weird (mixing numbers and dimnames) and it was the only op in PyTorch that had it. Keeping the support for this would have lead to an unnecessarily complicated code, so we dedided to remove it. Just add it to the relevant skip list.
> 
> Fwiw, @ezyang has an open PR trying to remove all dimnames from PyTorch (I really hope he manages to get it merged), so removing thisbparticularly odd overload is not that big a deal IMO

Seems the relevant list is `ALLOW_LIST` in `test/forward_backward_compatibility/check_forward_backward_compatibility.py`, does that seem right?

ngimel commented:
Yes, that's the list the op needs to be added to

Commit:
Update on "Add torch.unflatten and improve its docs"


    unflatten now has a free function version in torch.flatten in addition to
    the method in torch.Tensor.flatten.

    Updated docs to reflect this and polished them a little.
    For consistency, changed the signature of the int version of unflatten in
    native_functions.yaml.

    Some override tests were failing because unflatten has unusual
    characteristics in terms of the .int and .Dimname versions having
    different number of arguments so this required some changes
    to test/test_override.py

    Removed support for using mix of integer and string arguments
    when specifying dimensions in unflatten.

[ghstack-poisoned]
